### DISCUSSION TITLE
Exporting IR Data to any Arduino format only works 2 times?

### DISCUSSION THREAD

**Discussion by paulstary** [bengtmartensson/IrScrutinizer#502] [Export issues]
I have successfully imported a Global Cache IR command and converted it to Arduino Raw using the Export option. The resultant Sketch works great except it only allows me to send the IR command twice and then hangs somewhere in a library file. I have tried to look at these linked files but don't have the skill to understand what could be causing the hang.

Also, ALL of the (3) possible Arduino Export Sketches (Arduino Raw, Arduino/Infrared4Arduino and Arduino/Remote) have this same exact problem... only two sends and hang on both an Arduino Uno and Arduino Uno Every board.

Is this normal? I doubt it as the only way to use the Sketch for more than 2 sends is to reboot the Arduino.

IrScrutinizer has delivered on every step along my path except this final output result. Very frustrating! BTW, the IR command that is sent twice works perfectly! 

Does anyone have a clue what is happening here?

Thank you.

**Comment by bengtmartensson** [comment]
Hi,

first, please confirm that you are using the latest versions of IrScrutinizer (2.4.0 or the snapshot) and Infrared4Arduino (1.2.3, available in the Arduino library manager). Support for IRRemote is presently broken, see #444, so we leave that out until we have fixed that. Assuming that everything is the latest version, please post the broken export (preferably as gist).

**Comment by paulstary** [comment]
Thank you for the reply. I verified I am using IrScrutinizer 2.4.0 and Infrared4Arduino 1.2.3. I went through the process again and have posted the code in **[this gist](https://gist.github.com/paulstary/780e02f011fcb941c70f6563fd58aa45)** for you to look at.

Here is my process:
- Use IrS to Import the Control Tower IR code: TCL - TV - Roku TV Models - POWER ON
- Use IrS Export to default directory using Arduino Raw format by clicking on Export Rendered button
- This opens a Sketch in my Arduino v2.2.1 IDE named "arduino_raw_2023_11_01_08_57_07"
- I have an Arduino Uno R2 selected as the Board on COM 5
- I verify the Sketch compiles without errors
- I successfully upload the Sketch into the Arduino
- I have an IR emitter connected to Pin 3 of the Arduino
- I type "1" into the Serial Monitor command line when prompted to "Enter signal number"
- Then I type "1" when prompted to "Enter number of sends"
-  I verify that the IR command is sent by substituting a red LED. The IR command is sent
- The IR command also turns on the TCL TV in my lab when using the IR emitter
- I repeat this process a number of times but the MOST number of times I have been able to get it to work is 4
- It hangs with "Enter number of times to send it" on the Serial Monitor and typing any number into the command line does nothing after that
- I reset the Arduino and go through another round of tests which may result is 2-4 send cycles before it hangs again

**UPDATE**: In the course of gathering this data, the Uno board started working after an upload! I was able to execute 20+ cycles of IR commands without a hang. However, after connecting the Uno Every board to try it, I no longer can get the Uno to work. Back to hanging after 2-3 sends.

Thinking this might be a buffer issue, I manually edited the IR command data for the line below with a smaller data set:

`const microseconds_t intro_nec1_D0F0[] PROGMEM = { 9024U, 4512U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 1692U, 564U, 1692U, 564U, 1692U, 564U, 1692U, 564U, 1692U, 564U, 1692U, 564U, 1692U, 564U, 1692U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 1692U, 564U, 1692U, 564U, 1692U, 564U, 1692U, 564U, 1692U, 564U, 1692U, 564U, 1692U, 564U, 1692U, 564U, 39756 };`

I first tried to substitute the REPEAT data below which is short { 9024U, 2256U, 564U, 65535U } for the INTRO data. That works! No hang. Then I tried to edit the INTRO data and although some edits works (no hang), others do not. For example, `{ 9024U, 564U, 564U, 564U, 564U }` works while `{ 564U, 564U, 564U, 564U, 564U }` does not.

I hope with this information, you are able to resolve the problem. I will continue to explore on my end. This is the ultimate solution for my needs. It's very frustrating to be so close.

Thank you very much for your extensive, high-quality contributions in this area!

Paul Stary
Audio-Video Engineering

**Comment by bengtmartensson** [comment]
Thank you for the description. It looks quite OK. I'll  test if I can reproduce, but not today...

The canonical assumption is a memory leak.

**Comment by bengtmartensson** [comment]
I was able to reproduce your issue. 

However, earlier this year, I rewrote that export format, junking the combined IRremote/Infrared4Arduino version and replaced with a version for Infrared4Arduino only. This is not yet in a released version, but in the [snapshot](https://github.com/bengtmartensson/IrScrutinizer/releases/tag/ci-build). This appears to work fine. So, please download and try the snapshot version. (If you need the setup.exe version, just tell me.)

**Reply by paulstary** [reply]
This worked for my Uno board and I now have the solution I was looking for. Thank you very much!

I can't get it to work with the Arduino Nano Every board I want to use but I think this is just a pin definition issue. The Sketch compiles and uploads fine.

Thank you again for your fast response.
Paul Stary
